streams/src/test/java/org/apache/kafka/streams/integration/KStreamKStreamIntegrationTest.java

```diff
@@ -0,0 +1,186 @@
+        CONSUMER_CONFIG.put(ConsumerConfig.KEY_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+        CONSUMER_CONFIG.put(ConsumerConfig.VALUE_DESERIALIZER_CLASS_CONFIG, StringDeserializer.class);
+    }
+
+    @AfterAll
+    public static void closeCluster() {
+        CLUSTER.stop();
+    }
+
+    @BeforeEach
+    public void before(final TestInfo testInfo) throws IOException {
+        final String stateDirBasePath = TestUtils.tempDirectory().getPath();
+        final String safeTestName = safeUniqueTestName(getClass(), testInfo);
+        streamsConfig = getStreamsConfig(safeTestName);
+        streamsConfig.put(StreamsConfig.STATE_DIR_CONFIG, stateDirBasePath);
+    }
+
+    @AfterEach
+    public void after() throws IOException {
+        if (streams != null) {
+            streams.close();
+            streams = null;
+        }
+        IntegrationTestUtils.purgeLocalStreamsState(streamsConfig);
+    }
+
+    @Test
+    public void shouldOuterJoin() throws Exception {
+        final Set<KeyValue<String, String>> expectedOne = new HashSet<>();
+        expectedOne.add(new KeyValue<>("Key-1", "value1=left-1,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-2", "value1=left-2,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-3", "value1=left-3,value2=null"));
+        expectedOne.add(new KeyValue<>("Key-4", "value1=left-4,value2=null"));
+
+        verifyKStreamKStreamOuterJoin(expectedOne);
+    }
+
+    private void verifyKStreamKStreamOuterJoin(final Set<KeyValue<String, String>> expectedResult) throws Exception {
+        streams = prepareTopology(streamsConfig);
+
+        startApplicationAndWaitUntilRunning(Collections.singletonList(streams), ofSeconds(120));
+
+        PRODUCER_CONFIG.put(ProducerConfig.BOOTSTRAP_SERVERS_CONFIG, CLUSTER.bootstrapServers());
+        PRODUCER_CONFIG.put(ProducerConfig.ACKS_CONFIG, "all");
+        PRODUCER_CONFIG.put(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+        PRODUCER_CONFIG.put(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG, StringSerializer.class);
+
+        final List<KeyValue<String, String>> stream1 = asList(
+                new KeyValue<>("Key-1", "left-1"),
+                new KeyValue<>("Key-2", "left-2"),
+                new KeyValue<>("Key-3", "left-3"),
+                new KeyValue<>("Key-4", "left-4")
+        );
+
+        final List<KeyValue<String, String>> stream2 = asList(
+                new KeyValue<>("Key-1", "right-1"),
+                new KeyValue<>("Key-2", "right-2"),
+                new KeyValue<>("Key-3", "right-3"),
+                new KeyValue<>("Key-4", "right-4")
+        );
+
+        IntegrationTestUtils.produceKeyValuesSynchronously(STREAM_1, stream1, PRODUCER_CONFIG, MOCK_TIME);
+        MOCK_TIME.sleep(11000);
+        Thread.sleep(11000);
+        IntegrationTestUtils.produceKeyValuesSynchronously(STREAM_1, stream2, PRODUCER_CONFIG, MOCK_TIME);
```

Comment:
Is this supposed to be `STREAM_2` instead of `STREAM_1`?

Reply:
Actually no (cf the original description on the ticket). But I guess it does not matter to much anyway, and we could use both.

The point about using `STREAM_1` is, to make sure that the original keys are written twice into the same input topic partitions to advance time (given that for stream-stream join we have the "shared time tracker" it should not matter though.

Reply:
Maybe I should rename `stream1` and `stream2` in `firstInputBatch` and `secondInputBatch` or similar to decouple the same from "left" and "right" input, as both should go into the left input.